File: new_coder/51-60/BM55-没有重复项数字的全排列/main.java
import java.util.*;
public class Main {

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }

  public static class ListNode {
    int val;
    ListNode next = null;
    public ListNode(int val) { this.val = val; }
  }

  public void getNums(String s) {
    StringBuffer sb = new StringBuffer();
    for (int i = 0; i < s.length(); i++) {
      char ch = s.charAt(i);
      if (ch == '[') {
        sb.append('{');
      } else if (ch == ']') {
        sb.append('}');
      } else if (ch == '"') {
        sb.append("'");
      } else {
        sb.append(ch);
      }
    }
    System.out.println(sb.toString());
  }

  ArrayList<ArrayList<Integer>> ans = new ArrayList<>();
    int n;

    private void dfs(int[] num,boolean[] isVisited,ArrayList<Integer> list){
        if(list.size()==num.length){
            ans.add(new ArrayList<>(list));
            return;
        }
        for(int i=0;i<n;i++){
            if(!isVisited[i]){
                isVisited[i] = true;
                list.add(num[i]);
                dfs(num,isVisited,list);
                isVisited[i] = false;
                list.remove(list.size()-1);
            }
        }
    }
    public ArrayList<ArrayList<Integer>> permute (int[] num) {
        // write code here
        n = num.length;
        boolean[] isVisited = new boolean[n];
        dfs(num,isVisited,new ArrayList<>());
        return ans;
    }

  public static void main(String[] args) { Main main = new Main(); }
}

File: new_coder/51-60/BM55-没有重复项数字的全排列/main.cc
void dfs(vector<vector<int>>& arr,vector<int>& isVisited,vector<int>& tmp,vector<int>& num){
        if(tmp.size() == isVisited.size()){//思考边界
            arr.emplace_back(tmp);//push_back有没有拷贝构造，传进去的到底是真实的tmp还是tmp复制后的副本呢？
            return;
        }
        for(int i=0;i<isVisited.size();i++){
            if(isVisited[i] == 0){
                isVisited[i] = 1;
                tmp.push_back(num[i]);//添加到最后一位
                dfs(arr,isVisited,tmp,num);
                tmp.pop_back();//删除最后一个元素
                isVisited[i] = 0; //恢复现场
            }
        }
    }
    vector<vector<int>> permute(vector<int>& num) {
        // write code here
        vector<vector<int>> ans;
        int n = num.size();
        vector<int> isVisited(n,0);//如果isVisited[i] = 0 ,代表num[i]还没有被使用过
        vector<int> tmp;//最开始长度为0，存储每个排列，每找到一个合适的数就添加进去一个
        dfs(ans,isVisited,tmp,num);
        return ans;
    }